### ISSUE TITLE
Static upgrade instantiates upgraded services too soon - $injector not yet available

### ISSUE THREAD

**Issue by sebek64** [angular/angular#13811]
**I'm submitting a ...**
```
[X ] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

Factory for upgraded services is called before angular 1 is initialized - $injector is passed as undefined to the factory.

**Expected behavior**

Angular should instantiate services later when $injector is available.

**Minimal reproduction of the problem with instructions**

There are multiple scenarios which trigger this bug.

1. Service having ngOnInit lifecycle method. When you put it on a ng2 service that injects upgraded services, factory for this upgraded service is called sooner (during bootstrapModule, not in upgrade.bootstrap as it should, so $injector is not available).

https://plnkr.co/edit/ocxuE39B7Aci8wmSNEJJ

2. When we inject an upgraded service into a downgraded service and inject that downgraded service into a run() block, then the upgraded service is constructed with $injector still undefined.

https://plnkr.co/edit/eq8VXrAnMWExTDdZJEF6

3. When $http is used during an early phase (like with angular translate downloading translation files), then a service is also instantiated with $injector undefined.

config(function($httpProvider) { $httpProvider.interceptors.push(function(MyService) {...}) }));

**What is the motivation / use case for changing the behavior?**

When upgrading a slightly bigger project, it often contains a lot of config or run blocks. Sooner or later this issue is triggered.

**Please tell us about your environment:**

* **Angular version:** Tested both angular 2.4.1 and 4.0.0-beta.2, both are broken.

* **Browser:** Chrome 55 | Firefox 45 | maybe others

 
* **Language:** at least TypeScript 2.0 transpiled to ES5

* **Node (for AoT issues):** `node --version` =   6.9.3

**Comment by petebacondarwin** [comment]
I'll take a look early next week unless @gkalpak  gets to it first ;-)

**Comment by pdhar-tibco** [comment]
This happens also in an upgraded component too.

**Comment by bastijns-jeroen** [comment]
@petebacondarwin any news :-)

**Comment by petebacondarwin** [comment]
Sorry. It is on my todo list. Hopefully I will have some time later today.

**Comment by petebacondarwin** [comment]
As a starter I don't think that we will be able to support

1. Service having ngOnInit lifecycle method. When you put it on a ng2 service that injects upgraded services, factory for this upgraded service is called sooner (during bootstrapModule, not in upgrade.bootstrap as it should, so $injector is not available).

The design of upgrade is that the Angular module bootstraps before the AngularJS module, so if your Angular module provides services that have `ngOnInit` hooks then they will be called before we even begin to create the AngularJS injector. One workaround for this might be to lazy load the modules that have such services after the bootstrap of AngularJS has completed.

**Comment by petebacondarwin** [comment]
I think we ought to do better for scenario 2 - I'll look into it over the weekend.
I am not sure I understand scenario 3 - can you provide a plunker or more concrete reproduction steps?

**Comment by sebek64** [comment]
I think scenario 2 is basically the same as scenario 3 (i.e., by fixing 2 we get 3 fixed as well). I don't currently have access to the project that demonstrated this, so I wouldn't be probably able to provide a working plunker. I'm able to do this on the next Tuesday if that is necessary.

Impossibility of fixing 1: I think this should be mentioned in the documentation (https://angular.io/docs/ts/latest/guide/upgrade.html).

**Comment by petebacondarwin** [comment]
OK, so I think I know the cause of the problem.  If you look here: https://github.com/angular/angular/blob/master/modules/%40angular/upgrade/src/aot/upgrade_module.ts#L156

You can see that we create an AngularJS (1.x) upgrade module, which

a) depends upon the `modules` passed into the call to bootstrap
b) has a run block that initializes the upgraded injector service

The problem is that AngularJS will run the `run` blocks for dependent modules before the `run` blocks of the current module. So the injector has not been initialized by the time the modules' run blocks are executed.

I think we can solve this by creating two modules. A new one that simply initializes the injector and then the main upgrade module that puts that new module at the front of its dependencies list.

**Comment by petebacondarwin** [comment]
Please take a look at #14065 as a potential solution

**Comment by sebek64** [comment]
I can confirm that both 2 and 3 scenarios are fixed by #14065.

**Comment by sebek64** [comment]
And for scenario 1, I found that I'm not actually sure what is a correct application code (i.e., whether the provided plunker is a valid code). Lifecycle hooks are documented just for directives and components, but not for services. However, at least ngOnInit is called for services. Providing this method causes the service to be created sooner (and this lifecycle hook to be called during application bootstrap). When omitting the hook (and doing the same job in costructor, for instance), the service is instantiated later.

I don't see anywhere in the documentation whether a service should do its initialization job in constructor or in ngOnInit (for components, constructor approach is clearly a bad practice).
